// osmojs/main/codegen/osmosis/protorev/v1beta1/tx.registry.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.registry = exports.load = exports.MessageComposer = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _tx = require("./tx");
var registry = [["/osmosis.protorev.v1beta1.MsgSetHotRoutes", _tx.MsgSetHotRoutes], ["/osmosis.protorev.v1beta1.MsgSetDeveloperAccount", _tx.MsgSetDeveloperAccount]];
exports.registry = registry;
var load = function load(protoRegistry) {
  registry.forEach(function (_ref) {
    var _ref2 = (0, _slicedToArray2["default"])(_ref, 2),
      typeUrl = _ref2[0],
      mod = _ref2[1];
    protoRegistry.register(typeUrl, mod);
  });
};
exports.load = load;
var MessageComposer = {
  encoded: {
    setHotRoutes: function setHotRoutes(value) {
      return {
        typeUrl: "/osmosis.protorev.v1beta1.MsgSetHotRoutes",
        value: _tx.MsgSetHotRoutes.encode(value).finish()
      };
    },
    setDeveloperAccount: function setDeveloperAccount(value) {
      return {
        typeUrl: "/osmosis.protorev.v1beta1.MsgSetDeveloperAccount",
        value: _tx.MsgSetDeveloperAccount.encode(value).finish()
      };
    }
  },
  withTypeUrl: {
    setHotRoutes: function setHotRoutes(value) {
      return {
        typeUrl: "/osmosis.protorev.v1beta1.MsgSetHotRoutes",
        value: value
      };
    },
    setDeveloperAccount: function setDeveloperAccount(value) {
      return {
        typeUrl: "/osmosis.protorev.v1beta1.MsgSetDeveloperAccount",
        value: value
      };
    }
  },
  toJSON: {
    setHotRoutes: function setHotRoutes(value) {
      return {
        typeUrl: "/osmosis.protorev.v1beta1.MsgSetHotRoutes",
        value: _tx.MsgSetHotRoutes.toJSON(value)
      };
    },
    setDeveloperAccount: function setDeveloperAccount(value) {
      return {
        typeUrl: "/osmosis.protorev.v1beta1.MsgSetDeveloperAccount",
        value: _tx.MsgSetDeveloperAccount.toJSON(value)
      };
    }
  },
  fromJSON: {
    setHotRoutes: function setHotRoutes(value) {
      return {
        typeUrl: "/osmosis.protorev.v1beta1.MsgSetHotRoutes",
        value: _tx.MsgSetHotRoutes.fromJSON(value)
      };
    },
    setDeveloperAccount: function setDeveloperAccount(value) {
      return {
        typeUrl: "/osmosis.protorev.v1beta1.MsgSetDeveloperAccount",
        value: _tx.MsgSetDeveloperAccount.fromJSON(value)
      };
    }
  },
  fromPartial: {
    setHotRoutes: function setHotRoutes(value) {
      return {
        typeUrl: "/osmosis.protorev.v1beta1.MsgSetHotRoutes",
        value: _tx.MsgSetHotRoutes.fromPartial(value)
      };
    },
    setDeveloperAccount: function setDeveloperAccount(value) {
      return {
        typeUrl: "/osmosis.protorev.v1beta1.MsgSetDeveloperAccount",
        value: _tx.MsgSetDeveloperAccount.fromPartial(value)
      };
    }
  }
};
exports.MessageComposer = MessageComposer;
